File: backend/functions/src/features/authentication/domain/entities/user_entity.ts
import {z} from "zod";

import {onboardingStepSchema} from "./onboarding_step.js";
import {
  permissionPromptStateEntitySchema,
} from "./permission_prompt_state_entity.js";

/** Conta do usuário autenticado.
 *
 * [cpfFingerprint] existe no servidor (hash/HMAC); o cliente em geral não o
 * recebe.
 */
export const userEntitySchema = z.object({
  id: z.string().optional(),
  email: z.string(),
  fullName: z.string(),
  phoneNumber: z.string().nullable().optional(),
  onboardingStep: onboardingStepSchema,
  permissionPrompts: permissionPromptStateEntitySchema.default({
    notification: {seen: false, skipped: false},
    location: {seen: false, skipped: false},
    biometric: {seen: false, skipped: false},
  }),
  createdAt: z.coerce.date(),
  updatedAt: z.coerce.date(),
  cpf: z.string().optional(),
}).transform((row) => ({
  ...row,
  phoneNumber: row.phoneNumber ?? null,
}));

export type UserEntity = z.infer<typeof userEntitySchema>;

File: app/lib/features/users/domain/entities/user_entity.dart
import 'package:dart_mappable/dart_mappable.dart';
import '../../../../core/enums/authentication/onboarding_step.dart';
import '../../../permissions/domain/entities/permission_prompt_state_entity.dart';
part 'user_entity.mapper.dart';

/// Conta do usuário autenticado.
///
/// [cpfFingerprint] existe no servidor (hash/HMAC); o cliente em geral não o recebe.
@MappableClass()
class UserEntity with UserEntityMappable {
  UserEntity({
    this.id,
    required this.email,
    required this.phoneNumber,
    required this.fullName,
    required this.onboardingStep,
    this.permissionPrompts = const PermissionPromptStateEntity(),
    required this.createdAt,
    required this.updatedAt,
    this.cpf,
  });

  final String? id;
  final String email;
  final String phoneNumber;
  final String fullName;
  final OnboardingStep onboardingStep;
  final PermissionPromptStateEntity permissionPrompts;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? cpf;
}
